File: src/main/java/com/github/clevernucleus/playerex/PlayerEx.java
package com.github.clevernucleus.playerex;

import com.github.clevernucleus.dataattributes_dc.api.event.EntityAttributeModifiedEvents;
import com.github.clevernucleus.playerex.api.ExAPI;
import com.github.clevernucleus.playerex.api.event.LivingEntityEvents;
import com.github.clevernucleus.playerex.api.event.PlayerEntityEvents;
import com.github.clevernucleus.playerex.config.ConfigImpl;
import com.github.clevernucleus.playerex.factory.DamageFactory;
import com.github.clevernucleus.playerex.factory.EventFactory;
import com.github.clevernucleus.playerex.factory.ExScreenFactory;
import com.github.clevernucleus.playerex.factory.NetworkFactory;
import com.github.clevernucleus.playerex.factory.PlaceholderFactory;
import com.github.clevernucleus.playerex.factory.RefundFactory;
import com.github.clevernucleus.playerex.impl.CommandsImpl;

import eu.pb4.placeholders.api.Placeholders;
import me.shedaniel.autoconfig.AutoConfig;
import me.shedaniel.autoconfig.serializer.GsonConfigSerializer;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.command.v2.CommandRegistrationCallback;
import net.fabricmc.fabric.api.entity.event.v1.ServerPlayerEvents;
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents;
import net.fabricmc.fabric.api.networking.v1.ServerLoginConnectionEvents;
import net.fabricmc.fabric.api.networking.v1.ServerLoginNetworking;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.screen.ScreenHandlerType;
import net.minecraft.sound.SoundEvent;
import net.minecraft.util.Identifier;
import net.minecraft.registry.Registries;
import net.minecraft.registry.Registry;

public class PlayerEx implements ModInitializer {

    // Screen handler type for the custom screen
    public static final ScreenHandlerType<ExScreenFactory.Handler> EX_SCREEN = Registry
            .register(Registries.SCREEN_HANDLER, new Identifier(ExAPI.MODID, "ex_screen"), ExScreenFactory.type());

    // Sound events for level up and SP spend
    public static final SoundEvent LEVEL_UP_SOUND = SoundEvent.of(new Identifier(ExAPI.MODID, "level_up"));
    public static final SoundEvent SP_SPEND_SOUND = SoundEvent.of(new Identifier(ExAPI.MODID, "sp_spend"));

    @Override
    public void onInitialize() {
        // Register the configuration class
        AutoConfig.register(ConfigImpl.class, GsonConfigSerializer::new);

        // Register networking receivers for custom network packets
        ServerLoginNetworking.registerGlobalReceiver(NetworkFactory.CONFIG, NetworkFactory::loginQueryResponse);
        ServerPlayNetworking.registerGlobalReceiver(NetworkFactory.SCREEN, NetworkFactory::switchScreen);
        ServerPlayNetworking.registerGlobalReceiver(NetworkFactory.MODIFY, NetworkFactory::modifyAttributes);

        // Register custom commands
        CommandRegistrationCallback.EVENT.register(CommandsImpl::register);

        // Register event listeners for various server events
        ServerLoginConnectionEvents.QUERY_START.register(NetworkFactory::loginQueryStart);
        ServerLifecycleEvents.SERVER_STARTING.register(EventFactory::serverStarting);
        ServerPlayerEvents.COPY_FROM.register(EventFactory::reset);

        // Register LivingEntity events
        LivingEntityEvents.ON_HEAL.register(EventFactory::healed);
        LivingEntityEvents.EVERY_SECOND.register(EventFactory::healthRegeneration);
        LivingEntityEvents.ON_DAMAGE.register(EventFactory::onDamage);
        LivingEntityEvents.SHOULD_DAMAGE.register(EventFactory::shouldDamage);

        // Register PlayerEntity events
        PlayerEntityEvents.ON_CRIT.register(EventFactory::onCritAttack);
        PlayerEntityEvents.SHOULD_CRIT.register(EventFactory::attackIsCrit);

        // Register attribute modification events
        EntityAttributeModifiedEvents.CLAMPED.register(EventFactory::clamped);

        // Register damage modifications, refund conditions, and placeholders
        DamageFactory.STORE.forEach(ExAPI::registerDamageModification);
        RefundFactory.STORE.forEach(ExAPI::registerRefundCondition);
        PlaceholderFactory.STORE.forEach(Placeholders::register);

        // Register custom sound events
        Registry.register(Registries.SOUND_EVENT, LEVEL_UP_SOUND.getId(), LEVEL_UP_SOUND);
        Registry.register(Registries.SOUND_EVENT, SP_SPEND_SOUND.getId(), SP_SPEND_SOUND);
    }
}

File: src/main/kotlin/com/bibireden/playerex/PlayerEX.kt
package com.bibireden.playerex

import com.bibireden.data_attributes.api.DataAttributesAPI
import com.bibireden.data_attributes.api.attribute.EntityAttributeSupplier
import com.bibireden.data_attributes.api.event.EntityAttributeModifiedEvents
import com.bibireden.opc.api.OfflinePlayerCacheAPI
import com.bibireden.playerex.api.PlayerEXAPI
import com.bibireden.playerex.api.PlayerEXCachedKeys
import com.bibireden.playerex.api.PlayerEXCachedKeys.Level
import com.bibireden.playerex.api.attribute.PlayerEXAttributes
import com.bibireden.playerex.api.event.LivingEntityEvents
import com.bibireden.playerex.api.event.PlayerEXSoundEvents
import com.bibireden.playerex.api.event.PlayerEntityEvents
import com.bibireden.playerex.config.PlayerEXConfig
import com.bibireden.playerex.ext.component
import com.bibireden.playerex.factory.*
import com.bibireden.playerex.networking.NetworkingChannels
import com.bibireden.playerex.networking.NetworkingPackets
import com.bibireden.playerex.networking.registerServerbound
import com.bibireden.playerex.networking.types.UpdatePacketType
import de.dafuqs.additionalentityattributes.AdditionalEntityAttributes
import eu.pb4.placeholders.api.Placeholders
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.command.v2.CommandRegistrationCallback
import net.fabricmc.fabric.api.entity.event.v1.ServerPlayerEvents
import net.fabricmc.fabric.api.networking.v1.ServerLoginConnectionEvents
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.ai.attributes.Attributes
import org.slf4j.Logger
import org.slf4j.LoggerFactory

object PlayerEX : ModInitializer {
	const val MOD_ID: String = "playerex"

	@JvmField
	val LOGGER: Logger = LoggerFactory.getLogger(MOD_ID)

	@JvmField
	val CONFIG = PlayerEXConfig.createAndLoad()

	fun id(path: String) = ResourceLocation.tryBuild(MOD_ID, path)!!

	private val gimmick = listOf(
		"Let's do it right this time...",
		"We test in production (not really).",
	).random()

	override fun onInitialize() {
		NetworkingChannels.NOTIFICATIONS.registerClientboundDeferred(NetworkingPackets.Notify::class.java)

		NetworkingChannels.MODIFY.registerServerbound(NetworkingPackets.Update::class) { (type, id, amount), ctx ->
			val component = ctx.player.component
			EntityAttributeSupplier(id).get().ifPresent {
				when (type) {
					UpdatePacketType.Skill -> component.skillUp(it, amount)
					UpdatePacketType.Refund -> component.refund(it, amount)
				}
			}
		}
		NetworkingChannels.MODIFY.registerServerbound(NetworkingPackets.Level::class) { (amount), ctx ->
			ctx.player.component.levelUp(amount)
		}

		CommandRegistrationCallback.EVENT.register(PlayerEXCommands::register)

		ServerLoginConnectionEvents.QUERY_START.register(ServerNetworkingFactory::onLoginQueryStart)
		ServerPlayerEvents.COPY_FROM.register(EventFactory::reset)

		LivingEntityEvents.ON_HEAL.register(EventFactory::healed)
		LivingEntityEvents.ON_TICK.register(EventFactory::healthRegeneration)
		LivingEntityEvents.ON_DAMAGE.register(EventFactory::onDamage)
		LivingEntityEvents.SHOULD_DAMAGE.register(EventFactory::shouldDamage)

		PlayerEntityEvents.ON_CRITICAL.register(EventFactory::onCritAttack)
		PlayerEntityEvents.SHOULD_CRITICAL.register(EventFactory::attackIsCrit)

		DamageFactory.forEach(PlayerEXAPI::registerDamageModification)
		RefundFactory.forEach(PlayerEXAPI::registerRefundCondition)

		PlaceholderFactory.STORE.forEach(Placeholders::register)

		Registry.register(BuiltInRegistries.SOUND_EVENT, PlayerEXSoundEvents.LEVEL_UP_SOUND.location, PlayerEXSoundEvents.LEVEL_UP_SOUND)
		Registry.register(BuiltInRegistries.SOUND_EVENT, PlayerEXSoundEvents.SPEND_SOUND.location, PlayerEXSoundEvents.SPEND_SOUND)
		Registry.register(BuiltInRegistries.SOUND_EVENT, PlayerEXSoundEvents.REFUND_SOUND.location, PlayerEXSoundEvents.REFUND_SOUND)

		EntityAttributeModifiedEvents.MODIFIED.register { attribute, entity, _, _, _ ->
			if (entity?.level() == null) return@register // no entity & no world, skip

			if (!entity.level().isClientSide()) {
				if (attribute == Attributes.MAX_HEALTH) {
					entity.health = attribute.sanitizeValue(entity.health.toDouble()).toFloat()
				}
				else if (attribute == AdditionalEntityAttributes.LUNG_CAPACITY) {
					entity.airSupply = attribute.sanitizeValue(entity.airSupply.toDouble()).toInt()
				}
			}
		}

		LOGGER.info(gimmick)
	}

	init {
		OfflinePlayerCacheAPI.register(PlayerEXCachedKeys.LEVEL_KEY, Level::class.java, Level.CODEC) {
			Level(DataAttributesAPI.getValue(PlayerEXAttributes.LEVEL, it).map(Double::toInt).orElse(0))
		}
	}
}